File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3085_minimum-deletions-to-make-string-k-special/solutions/java/solution.java
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Collections;

class Solution {
    public int minimumDeletions(String word, int k) {
        HashMap<Character, Integer> freqMap = new HashMap<>();
        for (char ch : word.toCharArray()) {
            freqMap.put(ch, freqMap.getOrDefault(ch, 0) + 1);
        }
        
        ArrayList<Integer> frequencies = new ArrayList<>(freqMap.values());
        Collections.sort(frequencies);
        
        int totalChars = word.length();
        int minDeletions = Integer.MAX_VALUE;
        int n = frequencies.size();
        
        // Prefix sum to help calculate sum of frequencies efficiently
        int[] prefixSum = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            prefixSum[i] = prefixSum[i - 1] + frequencies.get(i - 1);
        }
        
        // Trying different frequency target from minimum to maximum frequency available
        for (int i = 0; i < n; i++) {
            int targetFreq = frequencies.get(i);
            int lo = 0, hi = i;
            while (lo < hi) {
                int mid = (lo + hi) / 2;
                if (targetFreq - frequencies.get(mid) > k) {
                    lo = mid + 1;
                } else {
                    hi = mid;
                }
            }
            
            int deleteBelow = prefixSum[i] - prefixSum[lo];
            int deleteAbove = 0;
            if (i + 1 < n) {
                deleteAbove = (prefixSum[n] - prefixSum[i + 1]) - (n - i - 1) * targetFreq;
            }
            
            int currentDeletion = deleteBelow + deleteAbove;
            minDeletions = Math.min(minDeletions, currentDeletion);
        }
        
        return minDeletions;
    }
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3085_minimum-deletions-to-make-string-k-special/solutions/python3/solution.py
class Solution:
    def minimumDeletions(self, word: str, k: int) -> int:
        from collections import Counter

        # Count frequencies of each character in the word
        freq = Counter(word)

        # Get the list of frequencies of each character
        counts = list(freq.values())

        # Number of characters to delete to make the word k-special
        to_delete = 0

        # Work with the sorted frequency list
        counts.sort()
        n = len(counts)

        # Use a two-pointer approach to find the minimal deletions
        left = 0
        right = 0
        min_deletions = float('inf')
        current_sum = 0

        # Move the right pointer to expand the window
        while right < n:
            while right < n and (right == left or counts[right] - counts[left] <= k):
                current_sum += counts[right]
                right += 1
            # Calculate the number of deletions for the current valid range
            # Target frequency is based on the smallest frequency in the window (counts[left])
            # All frequencies should be between counts[left] and counts[left] + k
            max_valid_freq = counts[left] + k
            target_sum = (right - left) * max_valid_freq
            min_deletions = min(min_deletions, current_sum - target_sum)
            
            # Slide the left pointer to reduce the window size
            current_sum -= counts[left]
            left += 1
        
        return sum(counts) - (current_sum + min_deletions)
